File: JavaScript/parser.js
class Parser {
  constructor(p){
    this.parse = p;
  }

  map(fn){
    return new Parser((inp) => this.parse(inp).map(([a, s]) => [fn(a), s]));
  }

  flatMap(fn){
    return new Parser((inp) => this.parse(inp).flatMap(([a, s]) => fn(a).parse(s)));
  }
  
  skip(fp){
    return this.flatMap((_) => fp());
  }
  
  orElseGet(fp){
    return new Parser((inp) => this.parse(inp).orElseGet(() => fp().parse(inp)));
  }

  orElse(p){
    return this.orElseGet(() => p);
  }

  ap(fp){
    return this.flatMap((f) => fp().map(f));
  }
  
  static pure(x){
    return new Parser((inp) => Maybe.Just([x, inp]));
  }

  static empty(){
    return new Parser((_) => Maybe.Nothing());
  }

  some(){
    return this.map((c) => (s) => c + s).ap(() => this.many());
  }
  
  many(){
    return this.some().orElse(empty_string);
  }
  
  token(){
    return between(spaces, spaces, () => this);
  }

  static rest(fp, ff, op, a){
    return op.flatMap((f) => fp().flatMap((b) => ff(f(a, b))))
      .orElseGet(() => Parser.pure(a));
  }
  
  rest_l(op, a){
    return Parser.rest(() => this, (b) => this.rest_l(op, b), op, a);
  }

  rest_r(op, a){
    return Parser.rest(() => this.chainr1(op), Parser.pure, op, a);
  }
  
  chainl1(op, negate_first){
    return this.flatMap((a) => this.rest_l(op, negate_first ? -a : a));
  }

  chainr1(op){
    return this.flatMap((a) => this.rest_r(op, a));
  }
}

let empty_string = Parser.pure("");

let anyChar = new Parser((inp) => inp.length === 0 ? Maybe.Nothing() : Maybe.Just([inp[0], inp.substring(1)]));

let satisfy = (pred) => anyChar.flatMap((c) => pred(c) ? Parser.pure(c) : Parser.empty());

let isWhitespace = (c) => /\s/.test(c);

let isDigit = (c) => /\d/.test(c);

let spaces = satisfy(isWhitespace).many();

let between = (open, close, fp) => open.skip(fp).flatMap((x) => close.skip(() => Parser.pure(x)));

File: Perl/src/parser.pl
package Parser;

require "./src/maybe.pl";
require "./src/pair.pl";

use strict;
use warnings;

use overload
  '|' => \&orElse,
  '>>' => \&skip,
  '~' => \&optional_s,
  fallback => 1; # Allows fallback to default behavior for unhandled ops

sub new {
  my ($class, $unp) = @_;
  bless { unp => $unp, type => 'Parser' }, $class
}

sub parse {
  my ($self, $inp) = @_;
  $self->{unp}($inp)
}

sub fmap {
  my ($self, $func) = @_;
  $self->andThen(sub { Parser->pure($func->(shift)) } )
}

sub andThen {
  my ($self, $func) = @_;
  Parser->new(sub {
    $self->parse(shift)->andThen(sub {
      my $pair = shift;
      $func->($pair->{first})->parse($pair->{second})
    })
  })
}

sub skip {
  my ($self, $fp) = @_;
  $self->andThen(sub { $fp->() })
}

sub orElse {
  my ($self, $fp) = @_;
  Parser->new(sub {
    my $inp = shift;
    $self->parse($inp)->orElse(sub {
        $fp->()->parse($inp)
    })
  })
}

sub orElseVal {
  my ($self, $other) = @_;
  $self->orElse(sub { $other })
}

sub apply {
  my ($self, $fp) = @_;
  $self->andThen(sub {
    $fp->()->fmap(shift)
  })
}

sub pure {
  my ($class, $x) = @_;
  Parser->new(sub {
    Maybe->Just(Pair->new($x, shift))
  })
}

sub empty {
  Parser->new(sub { Maybe->Nothing })
}

sub empty_string {
  Parser->pure("")
}

sub some {
  my $self = shift;
  $self->fmap(sub {
    my $c = shift;
    sub { $c . shift }
  })->apply(sub { $self->many })
}

sub many {
  shift->some->orElse(sub { empty_string })
}

sub any_char {
  Parser->new(sub {
    my $inp = shift;
    length($inp) > 0 ? Maybe->Just(Pair->new(substr($inp, 0, 1), substr($inp, 1))) : Maybe->Nothing
  })
}

sub satisfy {
  my ($class, $pred) = @_;
  any_char->andThen(sub {
    my $c = shift;
    $pred->($c) ? Parser->pure($c) : Parser->empty
  })
}

sub spaces {
  Parser->satisfy(sub {
    shift =~ /^\s*$/
  })->many
}
  
sub between {
  my ($class, $open, $close, $fp) = @_;
  $open >> sub {
    $fp->()->andThen(sub {
      my $x = shift;
      $close >> sub { Parser->pure($x) }
    })
  }
}

sub token {
  my $self = shift;
  Parser->between(spaces, spaces, sub { $self })
}

sub optional_s {
  shift | sub { empty_string }
}

sub rest {
  my ($fp, $ff, $op, $x) = @_;
  $op->andThen(sub {
    my $f = shift;
    $fp->()->andThen(sub {
      $ff->($f->($x, shift))
    })
  }) | sub { Parser->pure($x) }
}

sub rest_l {
  my ($self, $op, $x) = @_;
  rest(sub { $self }, sub { $self->rest_l($op, shift) }, $op, $x)
}

sub chainl1 {
  my ($self, $op, $negate_first) = @_;
  $self->andThen(sub { my $x = shift; $self->rest_l($op, $negate_first ? -$x : $x) })
}

sub rest_r {
  my ($self, $op, $x) = @_;
  rest(sub { $self->chainr1($op) }, sub { Parser->pure(shift) }, $op, $x)
}

sub chainr1 {
  my ($self, $op) = @_;
  $self->andThen(sub { $self->rest_r($op, shift) })
}
